crates/bevy_render/src/renderer/mod.rs

```diff
@@ -57,23 +57,95 @@ pub fn render_system(world: &mut World) {
 pub async fn initialize_renderer(
-    instance: &Instance,
+    #[cfg_attr(not(target_arch = "wasm32"), allow(unused))] app: &mut bevy_app::App,
+    instance: &wgpu::Instance,
+    #[cfg_attr(target_arch = "wasm32", allow(unused))] backends: wgpu::Backends,
     options: &mut WgpuOptions,
-    request_adapter_options: &RequestAdapterOptions<'_>,
-) -> (RenderDevice, RenderQueue) {
-    let adapter = instance
-        .request_adapter(request_adapter_options)
-        .await
-        .expect("Unable to find a GPU! Make sure you have installed required drivers!");
+) -> (RenderDevice, RenderAdapter, RenderQueue) {
+    let adapter = {
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            use crate::texture::{BevyDefault as _, DEFAULT_DEPTH_FORMAT};
+            let mut adapters: Vec<wgpu::Adapter> = instance.enumerate_adapters(backends).collect();
+            let (mut integrated, mut discrete, mut virt, mut cpu, mut other) =
+                (None, None, None, None, None);
 
-    info!("{:?}", adapter.get_info());
+            for (i, adapter) in adapters.iter().enumerate() {
+                let default_texture_format_features =
+                    adapter.get_texture_format_features(wgpu::TextureFormat::bevy_default());
+                let default_depth_format_features =
+                    adapter.get_texture_format_features(DEFAULT_DEPTH_FORMAT);
+                if default_texture_format_features
+                    .allowed_usages
+                    .contains(wgpu::TextureUsages::RENDER_ATTACHMENT)
+                    && default_depth_format_features
+                        .allowed_usages
+                        .contains(wgpu::TextureUsages::RENDER_ATTACHMENT)
+                {
+                    let info = adapter.get_info();
+                    match info.device_type {
+                        wgpu::DeviceType::IntegratedGpu => {
+                            integrated = integrated.or(Some(i));
+                        }
+                        wgpu::DeviceType::DiscreteGpu => {
+                            discrete = discrete.or(Some(i));
+                        }
+                        wgpu::DeviceType::VirtualGpu => {
+                            virt = virt.or(Some(i));
+                        }
+                        wgpu::DeviceType::Cpu => {
+                            cpu = cpu.or(Some(i));
+                        }
+                        wgpu::DeviceType::Other => {
+                            other = other.or(Some(i));
+                        }
+                    }
+                }
+            }
+
+            let preferred_gpu_index = match options.power_preference {
+                wgpu::PowerPreference::LowPower => {
+                    integrated.or(other).or(discrete).or(virt).or(cpu)
+                }
+                wgpu::PowerPreference::HighPerformance => {
+                    discrete.or(other).or(integrated).or(virt).or(cpu)
+                }
+            }
```

Comment:
this feels like recoding what https://docs.rs/wgpu/0.12.0/wgpu/struct.Instance.html#method.request_adapter should be doing

Reply:
Yep, it's a copy-pasted code from `wgpu`. But we should enumerate adapters here to make sure that we can use default texture format and default depth format as a render target for the selected adapter (for example, on my Android device Vulkan adapter can't use Depth32Float as a render target). `request adapter`  just return first adapter which meets requirements of `RequestAdapterOptions`.

Reply:
that makes me sad... I think we should either:
* switch from `Depth32Float` to `Depth24Plus` as the comment suggest, and which should be supported on Vulkan android
* upstream to wgpu a way to select an adapter based on the texture format it supports

Reply:
> switch from Depth32Float to Depth24Plus as the comment suggest, and which should be supported on Vulkan android (maybe with a cfg to check target architecture?)

As far as I know, only `VK_FORMAT_D16_UNORM`  supported across all Vulkan powered devices (99% based on gpuinfo https://vulkan.gpuinfo.org/listoptimaltilingformats.php?platform=android)

> upstream to wgpu a way to select an adapter based on the texture format it supports

`request_adapter` is implemented following by [WebGPU spec](https://gpuweb.github.io/gpuweb/#dom-gpu-requestadapter).  I don't think we can change it. 

That's why `wgpu-rs` exposed `enumerate_adapters` function. It's allowed users to select adapter based on their options. Perhaps in the future, we will need to check if the adapter supports any other features.

Reply:
Should we really be picking devices based on what texture / depth formats they support? That seems like a pretty trivial criteria relative to things like gpu features and performance. Wouldn't it make more sense to do runtime queries of supported formats and then pick the best one? Every device will support "something".

Reply:
@cart I implemented this behavior because bevy uses `Depth32Float` as a default depth format and it's required for `bevy_renderer`. When I tested this on a real device, it have `Depth32Float` format on OpenGL adapter, but Vulkan adapter doesn't support it. (Android 8.0).

We can remove this part from PR. It unblocks some Android devices (maybe most of the devices, but definitely not all of them).

Reply:
Yeah short term if we can move this forward without this behavior I think we should. I do acknowledge that this is a problem that needs solving, but I think it should be solved via querying supported depth formats at runtime _after_ we've selected our device based on more substantial criteria.

Anything that currently consumes hard-coded Depth32Float constants could call something like `RenderDevice::get_preferred_depth_format()` instead.

Reply:
> Anything that currently consumes hard-coded Depth32Float constants could call something like RenderDevice::get_preferred_depth_format() instead.

Sorry, maybe I confused you. We can query depth format only from **adapter** (correct me, if I'am wrong). For example, on real  device:
1. Create instance with Vulkan and GL backend
2. Enumerate adapters:
    1. Vulkan (Depth32Float is **not** available)
    2. OpenGL (Depth32Float is available)

What we should do in this case?

Reply:
We would choose vulkan (if that is what wgpu recommends based on the wgpu config) and pick a different depth format supported by the device.